tlf_journal: cache unflushed paths and update them on puts and flushes

strib opened the pull request:
This PR calculates the unflushed paths the first time they are asked for (even if multiple goroutines ask for them concurrently), and then keeps them up-to-date as puts and flushes happen.

Issue: KBFS-1476
Issue: KBFS-1477

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -183,6 +183,16 @@ type tlfJournal struct {
+	// `journalLock`.
+	unflushedPaths  map[MetadataRevision]map[string]bool
+	unflushedReady  chan struct{}
+	chainsPopulator chainsPathPopulator
```
Maybe this should just be passed in on construction time

strib replied:
I thought about passing this in at construction time, but journals are enabled via `libfs/journal_control_file.go`, and they don't have access to a `folderBlockOps` at that point.  Do you think we should change it so that enabling the journal goes through `folderBranchOps`?

akalin-keybase replied:
Ah, you're right...lemme think about it a bit more.

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -1238,11 +1360,90 @@ func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
+	// If someone initialized j.unflushedPaths, then try prepping the
+	// unflushed paths one more time.
+	if err == errRetryUnflushedPaths {
+		err := j.maybeWaitForUnflushedReady(ctx)
```
Okay, so I think I understand how this all works. But it's pretty complicated, and I think it can be simpler. Also, I think we can avoid blocking the `putMD` path on initializing the unflushed paths, and the retry logic.

So I think we need a separate object, say `unflushedPathCache` that holds:

- a lock, protecting all variables.
- a state enum, which is either uninitialized, initializing, or initialized.
- a channel (closed when the state goes to initialized)
- unflushedPaths and chainsPopulator, non-nil only when the state is initialized.
- a queue of MD objects / revisions, non-empty only when the state is initializing.

Let `unflushedPathCache` have a `getUnflushedPaths`, which will (under [r-]lock) immediately return `unflushedPaths` if initialized, but return an error otherwise.

Also have a `startInitializeOrWait` method, which will:

1) take the lock, defer unlock
2) If the state is uninitialized, transition to initializing, initialize the channel, and return true. If the state is initialized, return false.
3) Otherwise, release the lock and block on the channel. When unblocked, grab the lock and do step 2 again (which should work this time).

Also have an `abortInitialization` method, which will transition from initializing to uninitialized and close the channel.

Have an `appendToCache` method, which takes an `ImmutableRootMetadata`. If the state is uninitialized, it does nothing. If it is initializing, it will append the MD to the queue. If it is initialized, it will process with the chainsPopulator and append to `unflushedPaths`. Have a similar `removeFromCache` method.

Have a `finishInitialization` method, which takes a set of unflushed paths. If state is initializing and the queue is empty, set the internal unflushed paths to the given one, close the channel, and return success. Otherwise, clear the queue and return it to the caller to process.

Then `getJournalStatusWithPaths` will do something like:

1) Call `getUnflushedPaths`. If it returns something, we're done. Otherwise, call `startInitializeOrWait`. Then call `startInitializeOrWait`. If false is returned, call `getUnflushedPaths`, which should succeed. Otherwise, defer `abortInitializiation` on error.
3) Initialize unflushed paths with the results from `getJournalStatusWithRange`.
4) Call `finishInitialization`. If it succeeds, we're done. Otherwise, if it returns more ops, process them, and repeat until done.

Then the `putMD` path will just have to call `appendToCache`, and likewise with the flush path.

What do you think?

strib replied:
Minor issue: If we don't call `startInitializeOrWait` under `tlfJournal.journalLock` (and we shouldn't), then `putMD` could end up adding a revision to the queue that's included in the `getJournalStatusWithRange` results.  I guess that's not a big deal, `finishInitialization` can just ignore a queued revision if it's already been processed.

Also, the flush path can't just call `appendToCache`, there will probably need to be a separate queue for revision numbers that need to be removed.

strib replied:
(Oops, hit "Comment' too early above.)

Ah, I see now you mentioned a `removeFromCache` method.

Otherwise, I don't see any obvious concurrency differences with your model,so I'll give it a try.  Thanks for the suggestions.

strib commented:
@akalin-keybase: refactored based on your suggestions, ready for another look.

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -1219,27 +1271,77 @@ func (j *tlfJournal) getMDRange(
+	// somehow be sure the cache won't be initialized by the time we
+	// finish this put.
+	irmd := MakeImmutableRootMetadata(rmd, fakeMdID(1), time.Now())
+	perRevMap, err := j.unflushedPaths.prepUnflushedPaths(
```
So part of my suggestion was to remove this prep step entirely -- why can't all that logic be moved into `appendToCache`?

akalin-keybase replied:
But now that the path cache has its own lock, what could go wrong with prepping and appending after `putMD`? What "journal range" has to be updated atomically?

Something like this:

```
func putMD(...) {
  // grabs and releases lock.
  irmd, ... := doPutMD(...)
  j.unflushedPaths.prepAndAppend(irmd)
}
```

akalin-keybase replied:
ah, the uid and key? hmm...

strib replied:
No, if something calls `GetJournalStatusAndRange`, the `jStatus.RevisionStart` and `jStatus.RevisionEnd` need to match the unflushed paths returned by that function.

akalin-keybase replied:
Ah, ok, I think I see it now.

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -1219,27 +1271,77 @@ func (j *tlfJournal) getMDRange(
+	if !j.unflushedPaths.appendToCache(irmd, perRevMap) {
+		return MdID{}, true, nil
 	}
+	// TODO: remove the revision from the cache on any errors below?
```
Why not do `appendToCache` as the last thing?

strib replied:
Because we don't want to have the put succeed but need to retry the append due to a race, as we'll have to give up journal lock and leave things in an inconsistent state.

akalin-keybase commented on libkbfs/unflushed_path_cache.go:
```diff
@@ -0,0 +1,321 @@
+	// This loop is limited only by the lifetime of `ctx`.  TODO: we
+	// might want to add a way to block appendToCache calls if we loop
+	// too many times here waiting for an empty queue.
+	for {
```
All the unflushed path computation is purely informational, right? It seems to me that journal ops should take precedence over the unflushed path computation. What about returning partial results, since we already do that anyway? (And reverting back to uninitialized state.)

strib replied:
Yeah good point, returning the initial set but reverting back to uninitialized after a few attempts should work. We won't actually be returning any partial results, because all the caller of initialize cares about is the initial set. We will be just be doing a bunch of extra work on subsequent status calls to get the MD range and re-compute the paths again, but maybe that can't be helped.

strib commented:
Answered questions and addressed initialization starvation.  Ready for another look!

akalin-keybase commented on libkbfs/unflushed_path_cache.go:
```diff
@@ -0,0 +1,328 @@
+	}()
+
+	// The unflushed paths haven't been initialized yet.
+	if cpp == nil {
```
Okay, to remove the need for having to retry a prep/put, I guess passing in `chainsPopulator` into the constructor isn't that bad. I'm guessing have KBFSOps implement an (internal) interface with a `getChainsPathPopulator(TlfID)` method, and call that from `JournalServer.Enable`. What do you think?

strib replied:
Happy to change things around to pass in `cpp` (although the new auto-journaling stuff, by making new journals in `JournalServer.getTlfJournal`, significantly complicates that I think).  But I'm not sure it's a good idea to avoid the prep/put retry.  If we prepare before every put, even if the paths aren't yet cached, we are wasting a bunch of cycles/time for no reason, since the computation will just be thrown away.

Although, I guess if we did pass in `cpp` at construction, we could initialize the cache at `tlfJournal` construction time, and then we could get rid of the different states entirely, and always track the unflushed paths during the lifetime of the journal.  But that's a pretty major step that I'd rather do in a follow-up PR (especially since it involves solving the auto-journal question above).

strib replied:
On further thought, we probably can't get rid of the different states, because even if we try to populate the cache when creating the `tlfJournal`, it might be too large to do so quickly, and we'll have to return incomplete results until it is flushed enough to be populated quickly.  But I think it would still be a win (in a future PR) to be able to get `cpp` from the `KBFSOps` interface and at least try to initialize it at construction time.

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
 
 func (j *tlfJournal) getJournalStatusWithRange() (
-	TLFJournalStatus, []ImmutableBareRootMetadata, bool, error) {
+	TLFJournalStatus, []ImmutableBareRootMetadata,
```
The return behavior of this function deserves comment, I think. Something like:

```
getJournalStatusWithRange returns the journal status, and either a non-nil
unflushedPathsMap is returned, which can be used directly to fill in UnflushedPaths,
or a list of ImmutableBareRootMetadatas is returned (along with a bool indicating
whether that list is complete), which can be used to build an unflushedPathsMap.
```

akalin-keybase replied:
Should probably then name and put the args in this order: `unflushedPaths unflushedPathsMap, ibrmds []ImmutableRootMetadata, complete bool`

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
-		stop = jStatus.RevisionStart + 1000
-		complete = false
-	}
+	var ibrmds []ImmutableBareRootMetadata
```
Corresponding to the comment above, would be clearer like:

```
unflushedPaths := ...
if unflushedPaths != nil {
  return jStatus, unflushedPaths, nil, false, nil
}

if jStatus.RevisionEnd == MetadataRevisionUninitialized {
  return jStatus, nil, nil, true, nil
}

complete := true
...
return jStatus, nil, ibrmds, complete, nil
```

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
-	}
-	chains.mostRecentMD = irmds[len(irmds)-1]
+		// We need to make or initialize the unflushed paths.
+		if unflushedPaths == nil {
```
Seems clearer (and less-indented) to do:

```
// Use the cached paths.
if unflushedPaths != nil {
  break
}

// Otherwise, figure out the paths ... but don't cach it.
if !complete {
  ...
  break
}

// We need to...
doInit, err := ...
```

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
+
+			if doInit {
+				initSuccess := false
+				defer func() {
```
Don't you intend to scope this to the if clause? Do you need an anonymous function to wrap it? The error case will hit this, but not the `!initSuccess` case.

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
+					irmds)
+				if err != nil {
+					return TLFJournalStatus{}, err
+				}
```
Shouldn't you break if `initSuccess` is true?

strib replied:
It breaks regardless of the value of `initSuccess` -- there is a `break` below the `else` clause.  But I can re-work this to make it more clear.

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
+			} else {
+				j.log.CDebugf(ctx, "Waited for unflushed paths "+
+					"initialization, trying again to get the status")
+				continue
```
let fall through to loop back around?

strib commented:
Thanks for the review -- I addressed all comments except the discussion about `cpp`, which I plan to address in a followup PR, since it will be pretty majorly disruptive to this PR and will be easier to review separately.

strib commented:
I should add explicitly -- ready for another look, thanks!

akalin-keybase commented on libkbfs/tlf_journal.go:
```diff
@@ -1223,27 +1281,77 @@ func (j *tlfJournal) getMDRange(
-	MdID, error) {
+func (j *tlfJournal) doPutMD(ctx context.Context, rmd *RootMetadata,
+	irmd ImmutableRootMetadata, perRevMap unflushedPathsPerRevMap) (
+	MdID, bool, error) {
```
please name the return parameters (esp. the bool)

akalin-keybase commented on libkbfs/unflushed_path_cache.go:
```diff
@@ -0,0 +1,328 @@
+	}
+	cache := make(unflushedPathsMap)
+	// Only need to deep-copy the outer level of the map; the inner
+	// level (per-revision) is immutable.
```
Only if the callers don't modify it, right? May want to clarify / mention that in the function-level comment.

akalin-keybase commented on libkbfs/unflushed_path_cache.go:
```diff
@@ -0,0 +1,328 @@
+			len(newUnflushedPaths))
+	}
+
+	return newUnflushedPaths[irmd.Revision()], nil
```
if `addUnflushedPaths` somehow fills in the wrong key, this will silently return nil. Perhaps defensively check for that case and return an error/panic?

akalin-keybase commented on libkbfs/unflushed_path_cache.go:
```diff
@@ -0,0 +1,328 @@
+
+	initialUnflushedPaths := make(unflushedPathsMap)
+	// Only need to deep-copy the outer level of the map; the inner
+	// level (per-revision) is immutable.
```
same comment as above, perhaps "is immutable" -> "shouldn't be modified by the caller"
